mlir/lib/Conversion/GPUToLLVMSPV/GPUToLLVMSPV.cpp

```diff
@@ -239,23 +238,27 @@ struct GPUShuffleConversion final : ConvertOpToLLVMPattern<gpu::ShuffleOp> {
   }
 
   /// Get the subgroup size from the target or return a default.
-  static int getSubgroupSize(Operation *op) {
-    return spirv::lookupTargetEnvOrDefault(op)
-        .getResourceLimits()
-        .getSubgroupSize();
+  static std::optional<int> getSubgroupSize(Operation *op) {
+    auto constraints = op->getParentOfType<gpu::GPUModuleOp>().getConstraints();
+    return constraints ? constraints->getReqdSubgroupSize() : std::nullopt;
```

Comment:
`spirv::lookupTargetEnvOrDefault` (and `spirv::lookupTargetEnv`) searches for attributes in enclosing symbol tables, not only in `spirv.module` operations. Do we wanna do the same here? ([see](https://github.com/llvm/llvm-project/blob/8a43dc3efdd9bfba0bea32061ef2f3397a968eb9/mlir/lib/Dialect/SPIRV/IR/TargetAndABI.cpp#L184)).


Also, if there's no `gpu.module` parent, this will segfault.

Reply:
I've avoided the segfault, but it only makes sense to read the spatial extents from a `GPUModuleOp`.

Reply:
That can be an issue for projects like Triton not having `gpu.module` operations. Allowing that information in `module` would enable such projects.